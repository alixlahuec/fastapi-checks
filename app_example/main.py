from fastapi import FastAPI

from app_example.routers import resources


app = FastAPI(
    title="Test App",
    description="A sample application for testing",
)

app.include_router(resources.router)
