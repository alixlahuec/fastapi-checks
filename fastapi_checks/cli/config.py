from importlib import import_module
from pathlib import Path
from typing import Annotated, Optional

from fastapi import FastAPI
from pycosmiconfig import cosmiconfig
from typer import Context, Option

from fastapi_checks import models


explorer = cosmiconfig("fastapi_checks", package_prop="fastapi_checks")


def parse_entrypoint(loc: str) -> models.App:
    entrypoint = loc.split(":", maxsplit=1)

    try:
        assert len(entrypoint) == 2

        mod = import_module(entrypoint[0])
        app = getattr(mod, entrypoint[1])

        assert isinstance(app, FastAPI)

    except AssertionError:
        raise Exception("Location provided is not a FastAPI instance")

    return models.App(app=app)


# (Required) The FastAPI application to run checks against.
# The app's entrypoint can be specified in config or as a CLI argument.
App = Annotated[
    models.App,
    Option(
        parser=parse_entrypoint,
        help="Location of the FastAPI app.",
        show_default=False,
        hide_input=True,
    ),
]


def load_config(path: Optional[Path], ctx: Context) -> Optional[Path]:
    # If no config file is explicitly passed, look for one
    if path is None:
        try:
            autodetect_result = explorer.search()

            assert autodetect_result is not None
            path = autodetect_result.filepath

        except AssertionError:
            print("No config file provided, could not autodetect")
            return None

    assert isinstance(path, (Path, str))

    load_result = explorer.load(path)
    if load_result.is_empty:
        raise Exception(f"Config from {path} is empty")

    resolved_config = load_result.config
    assert isinstance(resolved_config, dict)

    # This default_map is not accessible through Context when running subcommands,
    # even though the defaults are interpreted correctly.
    # To make the values available to commands e.g. debug,
    # the dict is temporarily stored in the custom obj so that it can be passed to Globals()
    ctx.default_map = resolved_config
    ctx.obj = dict(config_values=resolved_config)

    return path


# (Optional) The path to a config file.
# If not provided, pycosmiconfig rules will be used to search for one.
# This must be parsed eagerly, to allow setting required arguments (like App) via config.
ConfigFile = Annotated[
    Optional[Path],
    Option(
        is_eager=True,
        show_default=False,
        callback=load_config,
        help="Path to a config file. If none is provided, auto-detection will be used.",
        dir_okay=False,
        file_okay=True,
        exists=True,
        readable=True,
    ),
]


# Callback for the app.
# The method's docstring is used as description for the app.
# The method's args are used to run setup operations.
def setup(
    ctx: Context,
    app: App,
    config: ConfigFile = None,
) -> None:
    """Run checks against a FastAPI application."""
    loaded_config = {}
    if hasattr(ctx, "obj"):
        loaded_config = (ctx.obj or {}).get("config_values", {})

    ctx.obj = models.Globals(app=app, config=config, config_values=loaded_config)
