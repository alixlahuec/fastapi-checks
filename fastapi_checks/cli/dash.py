from contextlib import contextmanager

import dash_bootstrap_components as dbc
from dash import Dash


@contextmanager
def run_dash(title: str):
    app = Dash(
        __name__,
        title=title,
        external_stylesheets=[
            dbc.themes.SUPERHERO,
            "https://cdn.jsdelivr.net/gh/AnnMarieW/dash-bootstrap-templates/dbc.min.css",
        ],
    )

    def render(children) -> None:
        app.layout = dbc.Container(
            children,
            className="dbc p-5",
        )

    yield render

    app.run(debug=True)
