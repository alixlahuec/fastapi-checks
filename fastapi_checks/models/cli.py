from pathlib import Path
from typing import Any, Optional

from pydantic import BaseModel

from fastapi_checks.models.app import App


class Globals(BaseModel):
    app: App
    config: Optional[Path] = None
    config_values: dict[str, Any] = dict()
