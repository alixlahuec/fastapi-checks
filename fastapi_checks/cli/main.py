from typer import Typer

from fastapi_checks.cli.commands import debug, routes
from fastapi_checks.cli.config import setup


app = Typer(
    name="fastapi_checks",
    callback=setup,
)

app.add_typer(debug.cli, name="debug")
app.add_typer(routes.cli, name="routes")


if __name__ == "__main__":  # pragma: no cover
    app()
