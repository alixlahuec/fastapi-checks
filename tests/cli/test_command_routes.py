from typer.testing import CliRunner

from fastapi_checks.cli.main import app


runner = CliRunner()


def test_routes_terminal(base_args, dash_run_mock, snapshot):
    res = runner.invoke(app, [*base_args, "routes", "--no-html"])

    assert res.exit_code == 0
    assert res.stdout == snapshot

    dash_run_mock.assert_not_called()


def test_routes_html(base_args, dash_run_mock, rich_print_mock):
    res = runner.invoke(app, [*base_args, "routes", "--html"])

    assert res.exit_code == 0

    dash_run_mock.assert_called_once_with(debug=True)

    rich_print_mock.assert_not_called()
