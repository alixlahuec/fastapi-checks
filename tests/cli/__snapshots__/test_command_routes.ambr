# serializer version: 1
# name: test_routes_terminal
  '''
                ╷        ╷             ╷             ╷             ╷              
                │        │             │             │ Dependency  │              
    Route       │ Method │ Description │ Dependenci… │ marks       │ Security     
  ╶─────────────┼────────┼─────────────┼─────────────┼─────────────┼─────────────╴
    /resource/… │ GET    │ Get         │ RouterDep   │ misc        │              
                │        │ resource    │ Config      │             │              
                │        │             │ some_depen… │             │              
                │        │             │ ClassDep    │             │              
                │        │             │ ClassDep    │             │              
    /resource/… │ GET    │ Get         │ RouterDep   │ misc        │ myapp:some…  
                │        │ resource    │ Config      │             │              
                │        │             │ ClassDep    │             │              
                │        │             │ some_depen… │             │              
                │        │             │ ClassDep    │             │              
    /resource/… │ DELETE │ Delete      │ RouterDep   │ tenant_enf… │              
                │        │ resource    │ Config      │             │              
                │        │             │ with_tenan… │             │              
                │        │             │ ClassDep    │             │              
                ╵        ╵             ╵             ╵             ╵              
  
  '''
# ---
